File: packages/frontend-agent/src/engine/gbnf.ts
import type { ToolSchema } from '../tools/types'

/**
 * Generate a GBNF grammar that constrains LFM2.5 tool-call decoding to the injected tool schema.
 * TS port of training/kbft/gbnf.py (kept in lockstep - validated against llama.cpp).
 *
 * The model supplies the POLICY (which tool, when, which id semantically); the grammar GUARANTEES the
 * STRUCTURE at decode time: only valid tool names, that tool's arg keys/types, enum values, and - when
 * `validIds` (the ids present in results so far) is given - the id argument is forced to one of those,
 * so id-grounding becomes unfaultable (a truncated/hallucinated id is impossible to emit). Applied via
 * wllama's `grammar` sampling param. `allowText` keeps a free-text reply branch so normal answers work.
 */

function lit(s: string): string {
  return '"' + s.replace(/\\/g, '\\\\').replace(/"/g, '\\"') + '"'
}
function sq(s: string): string {
  return lit("'" + s + "'")
}
function rn(name: string): string {
  // GBNF rule names take hyphens, not underscores
  return name.replace(/[^A-Za-z0-9]/g, '-')
}

function valueRule(
  tname: string,
  key: string,
  spec: Record<string, unknown>,
  validIds: string[] | undefined,
  idKeys: string[],
): string {
  if (idKeys.includes(key) && validIds && validIds.length) return `${rn(tname)}-${rn(key)}-id`
  const en = (spec as { enum?: unknown[] }).enum
  if (Array.isArray(en)) return '( ' + en.map((v) => sq(String(v))).join(' | ') + ' )'
  const t = (spec as { type?: string }).type
  if (t === 'integer' || t === 'number') return 'numval'
  if (t === 'boolean') return 'boolval'
  return 'strval'
}

export function buildToolGrammar(
  tools: ToolSchema[],
  validIds?: string[],
  idKeys: string[] = ['id'],
  allowText = true,
): string {
  const rules: string[] = ['sep ::= "," " "?']
  const callAlts: string[] = []
  const idAltsNeeded: [string, string][] = []

  for (const t of tools) {
    const name = t.name
    const props = (t.parameters?.properties ?? {}) as Record<string, Record<string, unknown>>
    const required = t.parameters?.required ?? []
    const r = rn(name)
    const reqFrags: string[] = []
    const optFrags: string[] = []
    for (const [key, spec] of Object.entries(props)) {
      const val = valueRule(name, key, spec, validIds, idKeys)
      if (idKeys.includes(key) && validIds && validIds.length) idAltsNeeded.push([name, key])
      const frag = `${r}-${rn(key)}`
      rules.push(`${frag} ::= ${lit(key + '=')} ${val}`)
      ;(required.includes(key) ? reqFrags : optFrags).push(frag)
    }
    // required args in order, comma-joined; optionals each appended optionally (matches the model)
    let seq = reqFrags.join(' sep ')
    for (const o of optFrags) seq = (seq ? seq + ' ' : '') + `( sep ${o} )?`
    rules.push(`${r}-call ::= ${lit(name + '(')} ${seq} ${lit(')')}`.replace(/ {2}/g, ' '))
    callAlts.push(`${r}-call`)
  }

  // One or more calls, comma-joined (`[fn1(...), fn2(...)]`); a single call is N=1. Mirrors kbft/gbnf.py.
  rules.push('call ::= ( ' + callAlts.join(' | ') + ' )')
  rules.push('toolcall ::= ' + lit('[') + ' call ( sep call )* ' + lit(']'))
  rules.push(
    'wrapped ::= ( ' + lit('<|tool_call_start|>') + ' )? toolcall ( ' + lit('<|tool_call_end|>') + ' )?',
  )
  for (const [name, key] of idAltsNeeded) {
    rules.push(`${rn(name)}-${rn(key)}-id ::= ` + (validIds as string[]).map(sq).join(' | '))
  }
  if (allowText) {
    // a free-text reply: first char isn't a tool-call opener, then any chars (incl. newlines)
    rules.push('reply ::= [^\\[<] [^\\x00]*')
    rules.push('root ::= wrapped | reply')
  } else {
    rules.push('root ::= wrapped')
  }
  rules.push(`strval ::= "'" [^']* "'"`)
  rules.push('numval ::= "-"? [0-9]+ ("." [0-9]+)?')
  rules.push('boolval ::= "True" | "False"')
  return rules.join('\n')
}

/** Collect string values under `id` keys from an arbitrary tool result (search results carry them). */
export function collectIds(value: unknown, into: Set<string>): void {
  if (Array.isArray(value)) {
    for (const v of value) collectIds(v, into)
  } else if (value && typeof value === 'object') {
    for (const [k, v] of Object.entries(value as Record<string, unknown>)) {
      if (k === 'id' && typeof v === 'string' && v) into.add(v)
      else collectIds(v, into)
    }
  }
}

File: training/kbft/gbnf.py
"""Generate a GBNF grammar that constrains LFM2.5 tool-call decoding to the injected tool schema.

Division of labour: the trained model supplies the POLICY (which tool, when, which id semantically);
this grammar GUARANTEES the STRUCTURE at decode time — only valid tool names, only that tool's arg
keys in the right shape, correct value types, enum values, and (when the caller passes the ids present
in the last search result) an id argument that can ONLY be one of those. That makes id-grounding
unfaultable — the `flux` -> `flux-machine` truncation class becomes impossible to emit.

The grammar is built PER TURN from the SAME aliased schema the system prompt injects, so it stays
retriever/site-agnostic (patterns, not semantics). Apply it via the llama.cpp / wllama `grammar`
sampling parameter.

Target format (LFM2.5 pythonic, single-quoted, optional special-token wrapper):
    <|tool_call_start|>[name(key='v', key2=123)]<|tool_call_end|>
"""

from __future__ import annotations

import re

_STRVAL = "strval ::= \"'\" [^']* \"'\""      # single-quoted string, no embedded quote (llama GBNF: no '.')
_NUMVAL = r'numval ::= "-"? [0-9]+ ("." [0-9]+)?'
_BOOLVAL = r'boolval ::= "True" | "False"'


def _lit(s: str) -> str:
    """GBNF double-quoted literal matching the exact string s."""
    return '"' + s.replace("\\", "\\\\").replace('"', '\\"') + '"'


def _sq(s: str) -> str:
    """GBNF literal matching the single-quoted pythonic string 'value'."""
    return _lit("'" + s + "'")


def _rn(name: str) -> str:
    """A safe GBNF rule-name fragment."""
    return re.sub(r"[^A-Za-z0-9]", "-", name)


def _value_rule(tname: str, key: str, spec: dict, valid_ids, id_keys) -> str:
    """Return the GBNF expression for one argument's value."""
    if key in id_keys and valid_ids:
        return f"{_rn(tname)}-{_rn(key)}-id"          # alternation of the real ids (defined later)
    if isinstance(spec.get("enum"), list):
        return "( " + " | ".join(_sq(str(v)) for v in spec["enum"]) + " )"
    t = spec.get("type")
    if t in ("integer", "number"):
        return "numval"
    if t == "boolean":
        return "boolval"
    return "strval"


def build_tool_grammar(tools: list[dict], valid_ids: list[str] | None = None,
                       id_keys=("id",), allow_text: bool = True) -> str:
    """GBNF constraining tool calls to `tools`. If `valid_ids` is given, any `id_keys` argument is
    constrained to exactly those ids. If `allow_text`, the model may instead emit a plain reply."""
    rules: list[str] = ['sep ::= "," " "?']
    call_alts, id_alts_needed = [], []
    for t in tools:
        fn = t.get("function", t)
        name = fn["name"]
        params = fn.get("parameters", {}) or {}
        props = params.get("properties", {}) or {}
        required = list(params.get("required", []))
        rn = _rn(name)
        req_frags, opt_frags = [], []
        for key, spec in props.items():
            val = _value_rule(name, key, spec, valid_ids, id_keys)
            if key in id_keys and valid_ids:
                id_alts_needed.append((name, key))
            frag = f"{rn}-{_rn(key)}"
            rules.append(f"{frag} ::= {_lit(key + '=')} {val}")
            (req_frags if key in required else opt_frags).append(frag)
        # required args in order, comma-joined; optionals each appended optionally (matches the model)
        seq = " sep ".join(req_frags)
        for o in opt_frags:
            seq = (seq + " " if seq else "") + f"( sep {o} )?"
        rules.append(f"{rn}-call ::= {_lit(name + '(')} {seq} {_lit(')')}".replace("  ", " "))
        call_alts.append(f"{rn}-call")
    # One or more calls, comma-joined inside the brackets (`[fn1(...), fn2(...)]`) — a single call is
    # just N=1. Lets the model batch parallel calls (e.g. two searches for a comparison) in one turn.
    rules.append("call ::= ( " + " | ".join(call_alts) + " )")
    rules.append("toolcall ::= " + _lit("[") + " call ( sep call )* " + _lit("]"))
    rules.append("wrapped ::= ( " + _lit("<|tool_call_start|>") + " )? toolcall ( "
                 + _lit("<|tool_call_end|>") + " )?")
    for name, key in id_alts_needed:
        rules.append(f"{_rn(name)}-{_rn(key)}-id ::= " + " | ".join(_sq(i) for i in valid_ids))
    if allow_text:
        # a free-text reply: first char isn't a tool-call opener, then any chars (incl. newlines)
        rules.append(r'reply ::= [^\[<] [^\x00]*')
        rules.append("root ::= wrapped | reply")
    else:
        rules.append("root ::= wrapped")
    rules += [_STRVAL, _NUMVAL, _BOOLVAL]
    return "\n".join(rules)
